### DISCUSSION TITLE
How to prevent Mirth adding HTML tags like <!DOCTYPE HTML> and <meta> to XML

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#5729] [General]
I created a HTTP listener channel that responds with simple, pure and very old-school HTML. My source transformer outbound message template is XML data type (for easy modification using E4X) and looks as follows:

```xml
<html>
  <head>
    <title>...</title>
  </head>
  <body>
    ...
  </body>
</html>
```

![image](https://user-images.githubusercontent.com/12202052/226732212-5bbb73dd-dfa6-4f38-95b2-01edeb356477.png)

Now, the problem is that Mirth adds some tags in the final outbound message:

```xml
<!DOCTYPE HTML><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>...
``` 

![image](https://user-images.githubusercontent.com/12202052/226732845-98a362a4-ca0e-4181-ba21-3a2cb73dd7bb.png)

... and this leads to a parsing error in the destination's inbound message ([link](https://github.com/nextgenhealthcare/connect/wiki/3.12.0---Upgrade-Guide#preventing-xml-external-entity-vulnerabilities)):

```
DETAILS:	TypeError: A DOCTYPE is not allowed in content.
```

I didn't add these tags to the `tmp` variable and I don't want Mirth to add something that I didn't define.

**So why is that? And how can I prevent Mirth doing this?**

**Comment by tonygermano** [comment]
I can confirm that this only happens in mirth, and not in vanilla Rhino. I suspect it has to do with this patch. https://github.com/nextgenhealthcare/connect/commit/1b423427bf73dedd378b139a911c065d435039f6

**Reply by tom08zehn** [reply]
Thx. Any idea how I can fix this within Mirth itself?

🤔 maybe a hacky workaround could be using `<some>` instead of `<html>` and replacing this in the response (not in `tmp`)

**Reply by tonygermano** [reply]
That could work, so that it doesn't interpret it as html. Coincidentally, this patch is also what breaks the native e4x pretty print option.

**Reply by tom08zehn** [reply]
Yes, I also had that issue with `XmlUtils.prettyPrint(xmlString)`

**Comment by jonbartels** [comment]
Does your use case require HTML or can you send XHTML? If you can send XHTML, do that since it *is* XML already.

**Reply by tom08zehn** [reply]
I could also use XHTML but I guess it wouldn't work either because it requires the DOCTYPE tag what makes it not parseable by Mirth's Rhino XML parser... would need to test but guess it's the same problem as the DOCTYPE has no closing element (therefore is not strict XML) and Rhino seems to disallow it as well.

https://www.w3schools.com/html/html_xhtml.asp

**Reply by tom08zehn** [reply]
Meanwhile, I solved it by creating the HTML response in the last destination (no more in the source transformer) so there's no next step where it needs to be parsed.

**Comment by jonbartels** [comment]
Oh, you're right. I think the last time I dealt with this issue my team had to get XHTML then treat it as text to add the correct XML header or doctype declarations.
